pkg/instance/instance.go

```diff
@@ -406,122 +413,187 @@ func (s *Server) spdkInstanceLog(req *rpc.InstanceLogRequest, srv rpc.InstanceSe
 	return grpcstatus.Error(grpccodes.Unimplemented, "spdk instance log is not supported")
 }
 
+func (s *Server) handleNotify(ctx context.Context, notifyChan chan struct{}, srv rpc.InstanceService_InstanceWatchServer) error {
+	logrus.Info("Start handling notify")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Info("Stopped handling notify")
+			return ctx.Err()
+		case <-notifyChan:
+			if err := srv.Send(&empty.Empty{}); err != nil {
+				return errors.Wrap(err, "failed to send instance response")
+			}
+		}
+	}
+}
+
 func (s *Server) InstanceWatch(req *empty.Empty, srv rpc.InstanceService_InstanceWatchServer) error {
 	logrus.Info("Start watching instances")
 
+	// Create a client for watching processes
+	pmClient, err := client.NewProcessManagerClient("tcp://"+s.processManagerServiceAddress, nil)
+	if err != nil {
+		return grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to create ProcessManagerClient").Error())
+	}
+	defer pmClient.Close()
+
+	var spdkClient *spdkclient.SPDKClient
+	if s.spdkEnabled {
+		// Create a client for watching SPDK engines and replicas
+		spdkClient, err = spdkclient.NewSPDKClient(s.spdkServiceAddress)
+		if err != nil {
+			return grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to create SPDK client").Error())
+		}
+		defer spdkClient.Close()
+	}
+
+	notifyChan := make(chan struct{})
+	defer close(notifyChan)
+
 	ctx := context.Background()
 	g, ctx := errgroup.WithContext(ctx)
 
 	g.Go(func() error {
-		return s.watchProcess(ctx, req, srv)
+		defer func() {
+			// Close the clients for closing streams and unblocking notifier Recv() with error.
+			pmClient.Close()
+			if s.spdkEnabled {
+				spdkClient.Close()
+			}
+		}()
+		return s.handleNotify(ctx, notifyChan, srv)
```

Comment:
I got what you mean. There are 2 minor issues:
1. This implementation avoids `notifier.Recv()` stuck but it cannot prevent `srv.Send` from blocking...
2. Closing clients twice looks a little bit ugly. Can we make this an unregistration goroutine that blindly closes/unregisters all entries of a sync.list? Then we can add the client to the list as long as we launch a new one.

Reply:
> This implementation avoids notifier.Recv() stuck but it cannot prevent srv.Send from blocking...

Any method for this..? It seems a common issue in a stream server?

Reply:
For 2., use a go routine for closing the clients.